src/pmi/src/pmi_wire.c

```diff
@@ -240,6 +240,26 @@ void PMII_pmi_init(struct PMII_pmi *pmi, int version, const char *cmd)
     pmi->num_tokens = 0;
 }
 
+void PMII_pmi_free_buf(struct PMII_pmi *pmi)
+{
+    MPL_free(pmi->buf);
+    pmi->buf = NULL;
+}
```

Comment:
The commit message says allocating the buffer supports reentrence, but are we really doing reentrent interfaces here or just multi-threaded? Using heap buffers helps us keep them thread private.

Reply:
Here by reentrant I mean the read routine can read another PMI message while a previous PMI message is still being handled. The previous PMI1 utility uses global static buffer. By allocating buffer with each PMI object, we allow multiple PMI buffers to be alive.

Reply:
Commit message revised:

```
    We need attach the buffer within each pmi object to allow multiple
    PMIU_cmd objects concurrently. Add PMIU_cmd_free_buf to free the buffer
    once done.
```